File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/StandardCalculator.kt
package com.qaantum.calculatorhub.calculators

import java.util.Stack

class StandardCalculator {
    private var input: String = "0"
    private var result: String = ""
    private var expression: String = ""

    fun getInput(): String = input
    fun getResult(): String = result
    fun getExpression(): String = expression

    fun onButtonPressed(buttonText: String) {
        when (buttonText) {
            "C" -> {
                input = "0"
                result = ""
                expression = ""
            }
            "⌫" -> {
                if (input.length > 1) {
                    input = input.substring(0, input.length - 1)
                } else {
                    input = "0"
                }
            }
            "=" -> {
                calculate()
            }
            else -> {
                if (input == "0" && !isOperator(buttonText)) {
                    input = buttonText
                } else {
                    input += buttonText
                }
            }
        }
    }

    private fun isOperator(x: String): Boolean {
        return x == "/" || x == "*" || x == "-" || x == "+"
    }

    private fun calculate() {
        try {
            var finalInput = input
            finalInput = finalInput.replace("×", "*")
            finalInput = finalInput.replace("÷", "/")

            val result = evaluateExpression(finalInput)
            expression = input
            
            // Remove decimal if integer
            val doubleResult = result.toDouble()
            if (doubleResult % 1 == 0.0) {
                this.result = doubleResult.toInt().toString()
                input = this.result
            } else {
                this.result = doubleResult.toString()
                input = this.result
            }
        } catch (e: Exception) {
            result = "Error"
        }
    }

    private fun evaluateExpression(expression: String): Double {
        val tokens = tokenize(expression)
        val postfix = infixToPostfix(tokens)
        return evaluatePostfix(postfix)
    }

    private fun tokenize(expression: String): List<String> {
        val tokens = mutableListOf<String>()
        var i = 0
        while (i < expression.length) {
            when {
                expression[i].isDigit() || expression[i] == '.' -> {
                    val sb = StringBuilder()
                    while (i < expression.length && (expression[i].isDigit() || expression[i] == '.')) {
                        sb.append(expression[i])
                        i++
                    }
                    tokens.add(sb.toString())
                    i--
                }
                expression[i] in "+-*/" -> {
                    tokens.add(expression[i].toString())
                }
            }
            i++
        }
        return tokens
    }

    private fun infixToPostfix(tokens: List<String>): List<String> {
        val output = mutableListOf<String>()
        val operators = Stack<String>()
        
        for (token in tokens) {
            when {
                token.matches(Regex("\\d+\\.?\\d*")) -> output.add(token)
                token == "(" -> operators.push(token)
                token == ")" -> {
                    while (operators.isNotEmpty() && operators.peek() != "(") {
                        output.add(operators.pop())
                    }
                    operators.pop()
                }
                else -> {
                    while (operators.isNotEmpty() && precedence(operators.peek()) >= precedence(token)) {
                        output.add(operators.pop())
                    }
                    operators.push(token)
                }
            }
        }
        
        while (operators.isNotEmpty()) {
            output.add(operators.pop())
        }
        
        return output
    }

    private fun precedence(op: String): Int {
        return when (op) {
            "+", "-" -> 1
            "*", "/" -> 2
            else -> 0
        }
    }

    private fun evaluatePostfix(postfix: List<String>): Double {
        val stack = Stack<Double>()
        
        for (token in postfix) {
            when {
                token.matches(Regex("\\d+\\.?\\d*")) -> stack.push(token.toDouble())
                token == "+" -> {
                    val b = stack.pop()
                    val a = stack.pop()
                    stack.push(a + b)
                }
                token == "-" -> {
                    val b = stack.pop()
                    val a = stack.pop()
                    stack.push(a - b)
                }
                token == "*" -> {
                    val b = stack.pop()
                    val a = stack.pop()
                    stack.push(a * b)
                }
                token == "/" -> {
                    val b = stack.pop()
                    val a = stack.pop()
                    stack.push(a / b)
                }
            }
        }
        
        return stack.pop()
    }
}

File: Calculator/ios-native/CalculatorHub/Calculators/StandardCalculator.swift
import Foundation

class StandardCalculator: ObservableObject {
    @Published var input: String = "0"
    @Published var result: String = ""
    @Published var expression: String = ""

    func onButtonPressed(_ buttonText: String) {
        switch buttonText {
        case "C":
            input = "0"
            result = ""
            expression = ""
        case "⌫":
            if input.count > 1 {
                input = String(input.dropLast())
            } else {
                input = "0"
            }
        case "=":
            calculate()
        default:
            if input == "0" && !isOperator(buttonText) {
                input = buttonText
            } else {
                input += buttonText
            }
        }
    }

    private func isOperator(_ x: String) -> Bool {
        return x == "/" || x == "*" || x == "-" || x == "+"
    }

    private func calculate() {
        do {
            var finalInput = input
            finalInput = finalInput.replacingOccurrences(of: "×", with: "*")
            finalInput = finalInput.replacingOccurrences(of: "÷", with: "/")

            let result = try evaluateExpression(finalInput)
            expression = input

            // Remove decimal if integer
            if result.truncatingRemainder(dividingBy: 1) == 0 {
                self.result = String(Int(result))
                input = self.result
            } else {
                self.result = String(result)
                input = self.result
            }
        } catch {
            result = "Error"
        }
    }

    private func evaluateExpression(_ expression: String) throws -> Double {
        let tokens = tokenize(expression)
        let postfix = try infixToPostfix(tokens)
        return try evaluatePostfix(postfix)
    }

    private func tokenize(_ expression: String) -> [String] {
        var tokens: [String] = []
        var i = expression.startIndex

        while i < expression.endIndex {
            if expression[i].isNumber || expression[i] == "." {
                var token = ""
                while i < expression.endIndex && (expression[i].isNumber || expression[i] == ".") {
                    token.append(expression[i])
                    i = expression.index(after: i)
                }
                tokens.append(token)
                if i < expression.endIndex {
                    i = expression.index(before: i)
                }
            } else if "+-*/".contains(expression[i]) {
                tokens.append(String(expression[i]))
            }
            i = expression.index(after: i)
        }
        return tokens
    }

    private func infixToPostfix(_ tokens: [String]) throws -> [String] {
        var output: [String] = []
        var operators: [String] = []

        for token in tokens {
            if Double(token) != nil {
                output.append(token)
            } else if token == "(" {
                operators.append(token)
            } else if token == ")" {
                while !operators.isEmpty && operators.last != "(" {
                    output.append(operators.removeLast())
                }
                if operators.isEmpty {
                    throw NSError(domain: "Invalid expression", code: 1)
                }
                operators.removeLast()
            } else {
                while !operators.isEmpty && precedence(operators.last!) >= precedence(token) {
                    output.append(operators.removeLast())
                }
                operators.append(token)
            }
        }

        while !operators.isEmpty {
            output.append(operators.removeLast())
        }

        return output
    }

    private func precedence(_ op: String) -> Int {
        switch op {
        case "+", "-": return 1
        case "*", "/": return 2
        default: return 0
        }
    }

    private func evaluatePostfix(_ postfix: [String]) throws -> Double {
        var stack: [Double] = []

        for token in postfix {
            if let number = Double(token) {
                stack.append(number)
            } else {
                guard stack.count >= 2 else {
                    throw NSError(domain: "Invalid expression", code: 1)
                }
                let b = stack.removeLast()
                let a = stack.removeLast()

                switch token {
                case "+": stack.append(a + b)
                case "-": stack.append(a - b)
                case "*": stack.append(a * b)
                case "/":
                    guard b != 0 else {
                        throw NSError(domain: "Division by zero", code: 1)
                    }
                    stack.append(a / b)
                default: throw NSError(domain: "Unknown operator", code: 1)
                }
            }
        }

        guard stack.count == 1 else {
            throw NSError(domain: "Invalid expression", code: 1)
        }

        return stack[0]
    }
}

